tests/manual_message.go

```diff
@@ -0,0 +1,151 @@
+	fundRelayerTxA := utils.CreateNativeTransferTransaction(
+		ctx, subnetAInfo, fundedKey, relayerAddress, fundAmount,
+	)
+	utils.SendTransactionAndWaitForSuccess(ctx, subnetAInfo, fundRelayerTxA)
+	fundRelayerTxB := utils.CreateNativeTransferTransaction(
+		ctx, subnetBInfo, fundedKey, relayerAddress, fundAmount,
+	)
+	utils.SendTransactionAndWaitForSuccess(ctx, subnetBInfo, fundRelayerTxB)
+
+	//
+	// Send two Teleporter message on Subnet A, before the relayer is running
+	//
+
+	log.Info("Sending two teleporter messages on subnet A")
+	// This message will be delivered by the relayer
+	receipt1, _, id1 := sendBasicTeleporterMessage(ctx, subnetAInfo, subnetBInfo, fundedKey, fundedAddress)
+	msg1 := getWarpMessageFromLog(ctx, receipt1, subnetAInfo)
+
+	// This message will not be delivered by the relayer
+	_, _, id2 := sendBasicTeleporterMessage(ctx, subnetAInfo, subnetBInfo, fundedKey, fundedAddress)
+
+	//
+	// Set up relayer config to deliver one of the two previously sent messages
+	//
+	relayerConfig := testUtils.CreateDefaultRelayerConfig(
+		subnetAInfo,
+		subnetBInfo,
+		teleporterContractAddress,
+		fundedAddress,
+		relayerKey,
+	)
+	relayerConfig.ManualWarpMessages = []config.ManualWarpMessage{
+		{
+			UnsignedMessageBytes:    hex.EncodeToString(msg1.Bytes()),
+			SourceBlockchainID:      subnetAInfo.BlockchainID.String(),
+			DestinationBlockchainID: subnetBInfo.BlockchainID.String(),
+			SourceAddress:           teleporterContractAddress.Hex(),
+			DestinationAddress:      teleporterContractAddress.Hex(),
+		},
+	}
+
+	relayerConfigPath := writeRelayerConfig(relayerConfig)
+
+	//
+	// Build Relayer
+	//
+	// Build the awm-relayer binary
+	cmd := exec.Command("./scripts/build.sh")
+	out, err := cmd.CombinedOutput()
+	fmt.Println(string(out))
+	Expect(err).Should(BeNil())
+
+	//
+	// Run the Relayer. On startup, we should deliver the message provided in the config
+	//
+
+	// Subscribe to the destination chain
+	newHeadsB := make(chan *types.Header, 10)
+	sub, err := subnetBInfo.WSClient.SubscribeNewHead(ctx, newHeadsB)
+	Expect(err).Should(BeNil())
+	defer sub.Unsubscribe()
+
+	log.Info("Starting the relayer")
+	relayerCmd, relayerCancel := testUtils.RunRelayerExecutable(ctx, relayerConfigPath)
+
+	log.Info("Waiting for a new block confirmation on subnet B")
+	<-newHeadsB
+	delivered1, err := subnetBInfo.TeleporterMessenger.MessageReceived(
+		&bind.CallOpts{}, id1,
+	)
+	Expect(err).Should(BeNil())
+	Expect(delivered1).Should(BeTrue())
+
+	log.Info("Waiting for 10s to ensure no new block confirmations on destination chain")
+	Consistently(newHeadsB, 10*time.Second, 500*time.Millisecond).ShouldNot(Receive())
+
+	delivered2, err := subnetBInfo.TeleporterMessenger.MessageReceived(
+		&bind.CallOpts{}, id2,
+	)
+	Expect(err).Should(BeNil())
+	Expect(delivered2).Should(BeFalse())
+
+	// Cancel the command and stop the relayer
+	relayerCancel()
+	_ = relayerCmd.Wait()
```

Comment:
It's not necessarily the case that we always want to kill the relayer at the end of the test case. For example in basic_relay.go, we kill the relayer in the middle of the test so that we can rewrite the config. I'm not sure how defer works if the variable containing the function is overwritten in between the defer statement and its execution, but I think explicitly calling it like we do here avoids such confusion, even if it's slightly more error prone.

In any case, if any test errors, the whole suite stops so there's no risk of a failed test impacting subsequent tests.

Reply:
> In any case, if any test errors, the whole suite stops so there's no risk of a failed test impacting subsequent tests.

I'm pretty sure that's not true. AFAIK this only happens if you pass `--fail-fast` to ginkgo.

> I'm not sure how defer works if the variable containing the function is overwritten

As long as you don't pass the cancelfunc as a parameter to the deferred function, it'll just grab the value from the scope when it executes. This means you'll get the most recent `relayerCancel`. Cancelfuncs are also idempotent so it's fine to defer as many of them as we want.

[Explanation of defer closures here ](https://stackoverflow.com/questions/16010694/how-golangs-defer-capture-closures-parameter)

I do really think we should defer the cancelfunc, but the `relayerCmd.Wait()` should be fine to call at the end.

Reply:
> I'm pretty sure that's not true. AFAIK this only happens if you pass --fail-fast to ginkgo.

Oh, yeah you're right. We should probably use `--fail-fast` since our tests share network state. Try as we might to isolate them, we're bound to make mistakes.

As for the cancelfunc, it still seems odd to me to defer it in some instances and call it directly in others. In the case of a mid-test relayer restart, we'd call the cancelfunc along with relayer.wait, but at the end of the test we'd only call relayer.wait. My preference is to keep the usage consistent in each test, so in that spirit I'll defer the call in manual_message.go but leave it as is in basic_relay.go

Reply:
Oh yeah we wouldn't be able to defer just the cancelfunc. I'm still going to suggest we defer both then, since that's idiomatically how to acheive RAII in go. Cancelfunc is idempotent, and relayerCmd.Wait() just returns an error if it's already been called, but we're not checking the error anyway.

Reply:
As far as --fail-fast, I'm easy either way, at least for now, but I don't think we should lean on it as an excuse not to clean up the state if we can

Reply:
Addressed in https://github.com/ava-labs/awm-relayer/pull/153/commits/853222c12a5f02f2edb26c35b656e982217743f8. Thanks for authoring that!